File: sdk/dotnet/src/DependencyInjection/ServicesProvider.fs
namespace StarFederation.Datastar.DependencyInjection

open System
open System.Runtime.CompilerServices
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.DependencyInjection
open StarFederation.Datastar

[<Extension>]
type ServiceCollectionExtensionMethods() =
    [<Extension>]
    static member AddDatastar (serviceCollection:IServiceCollection) =
        serviceCollection
            .AddHttpContextAccessor()
            .AddScoped<IDatastarSignalsReaderService>(fun (svcPvd:IServiceProvider) ->
                let httpContextAccessor = svcPvd.GetService<IHttpContextAccessor>()
                let signalsHttpHandler = SignalsHttpHandlers httpContextAccessor.HttpContext.Request
                SignalsReaderService signalsHttpHandler
                )
            .AddScoped<IDatastarServerSentEventService>(fun (svcPvd:IServiceProvider) ->
                let httpContextAccessor = svcPvd.GetService<IHttpContextAccessor>()
                let sseHttpHandler = ServerSentEventHttpHandlers httpContextAccessor.HttpContext.Response
                sseHttpHandler.StartResponse() |> ignore
                //sseHttpHandler.AddHeader ("X-Accel-Buffering", "no")  // an example of adding a header
                ServerSentEventService sseHttpHandler
                )

File: sdk/dotnet/csharp/src/DependencyInjection/ServicesProvider.cs
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Core = StarFederation.Datastar.FSharp;

namespace StarFederation.Datastar.DependencyInjection;

public static class ServiceCollectionExtensionMethods
{
    public static IServiceCollection AddDatastar(this IServiceCollection serviceCollection)
    {
        serviceCollection
            .AddHttpContextAccessor()
            .AddScoped<IDatastarService>(svcPvd =>
            {
                IHttpContextAccessor? httpContextAccessor = svcPvd.GetService<IHttpContextAccessor>();
                Core.ISendServerEvent sseHttpHandler = new Core.ServerSentEventHttpHandler(httpContextAccessor!.HttpContext!.Response);
                Core.IReadSignals signalsHttpHandler = new Core.SignalsHttpHandler(httpContextAccessor!.HttpContext!.Request);
                return new DatastarService(sseHttpHandler, signalsHttpHandler);
            });
        return serviceCollection;
    }
}